include/micm/solver/backward_euler.inl

```diff
@@ -0,0 +1,173 @@
+/* Copyright (C) 2023-2024 National Center for Atmospheric Research
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+enum class MicmBackwardEulerErrc
+{
+  SubStepTooSmall = 1,
+};
+
+namespace std
+{
+  template<>
+  struct is_error_condition_enum<MicmBackwardEulerErrc> : true_type
+  {
+  };
+}  // namespace std
+
+namespace
+{
+  class BackwardEulerErrorCategory : public std::error_category
+  {
+   public:
+    const char* name() const noexcept override
+    {
+      return "MICM BackwardEuler";
+    }
+    std::string message(int ev) const override
+    {
+      switch (static_cast<MicmBackwardEulerErrc>(ev))
+      {
+        case MicmBackwardEulerErrc::SubStepTooSmall:
+          return "Substep too small";
+        default: return "Unknown error";
+      }
+    }
+  };
+
+  const BackwardEulerErrorCategory backwardEulerErrorCategory{};
+}  // namespace
+
+inline std::error_code make_error_code(MicmBackwardEulerErrc e)
+{
+  return { static_cast<int>(e), backwardEulerErrorCategory };
+}
+
+namespace micm
+{
+  inline BackwardEuler::BackwardEuler()
+      : processes_()
+  {
+  }
+
+  inline BackwardEuler::BackwardEuler(
+        const System& system,
+        const std::vector<Process>& processes)
+      : processes_()
+  {
+  }
+
+  enum class LastTimeStepCut
+  {
+    None,
+    Half,
+    Tenth
+  };
+
+  inline void BackwardEuler::Solve(double time_step, auto state, auto linear_solver, auto process_set)
```

Comment:
Probably we can not do a strict apple-to-apple comparison here. But CAM-Chem calculates a chunk of data for the chemistry solver (https://github.com/ESCOMP/CAM/blob/cam_development/src/chemistry/pp_trop_strat_mam4_vbs/mo_imp_sol.F90#L198) rather than all the grid cells at once, which is done in MICM right now.

Reply:
We solve multiple grid cells all at once, just the same as our Rosenbrock implementation. If we need to solve chunks of grid cells, I think we should push to have that logic outside of the solve routine. The camchem implementation is complicated and the readability gained from not handling chunked solving I think is far more important than the small computational costs we'll have from stack frames added without a for loop on the outside of the solve function that loops over chunks